src/hasher512.cpp

```diff
@@ -0,0 +1,235 @@
+	{
+		static HCRYPTPROV prov = make_crypt_provider();
+		return prov;
+	}
+}
+#endif
+
+namespace libtorrent
+{
+	hasher512::hasher512()
+	{
+#ifdef TORRENT_USE_LIBGCRYPT
+		gcry_md_open(&m_context, GCRY_MD_SHA512, 0);
+#elif TORRENT_USE_COMMONCRYPTO
+		CC_SHA512_Init(&m_context);
+#elif TORRENT_USE_CRYPTOAPI
+		if (CryptCreateHash(get_crypt_provider(), CALG_SHA_512, 0, 0, &m_context) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+#elif defined TORRENT_USE_LIBCRYPTO
+		SHA512_Init(&m_context);
+#else
+		SHA512_init(&m_context);
+#endif
+	}
+
+	hasher512::hasher512(span<char const> data)
+		: hasher512()
+	{
+		update(data);
+	}
+
+	hasher512::hasher512(char const* data, int len)
+		: hasher512()
+	{
+		TORRENT_ASSERT(len > 0);
+		update({data, size_t(len)});
+	}
+
+#ifdef TORRENT_USE_LIBGCRYPT
+	hasher512::hasher512(hasher512 const& h)
+	{
+		gcry_md_copy(&m_context, h.m_context);
+	}
+
+	hasher512& hasher512::operator=(hasher512 const& h)
+	{
+		if (this == &h) return;
+		gcry_md_close(m_context);
+		gcry_md_copy(&m_context, h.m_context);
+		return *this;
+	}
+#elif TORRENT_USE_CRYPTOAPI
+	hasher512::hasher512(hasher512 const& h)
+	{
+		if (CryptDuplicateHash(h.m_context, 0, 0, &m_context) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+	}
+
+	hasher512& hasher512::operator=(hasher512 const& h)
+	{
+		if (this == &h) return *this;
+		CryptDestroyHash(m_context);
+		if (CryptDuplicateHash(h.m_context, 0, 0, &m_context) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+		return *this;
+	}
+#else
+	hasher512::hasher512(hasher512 const&) = default;
+	hasher512& hasher512::operator=(hasher512 const&) = default;
+#endif
+
+	hasher512& hasher512::update(char const* data, int len)
+	{
+		return update({data, size_t(len)});
+	}
+
+	hasher512& hasher512::update(span<char const> data)
+	{
+		TORRENT_ASSERT(!data.empty());
+#ifdef TORRENT_USE_LIBGCRYPT
+		gcry_md_write(m_context, data.data(), data.size());
+#elif TORRENT_USE_COMMONCRYPTO
+		CC_SHA512_Update(&m_context, reinterpret_cast<unsigned char const*>(data.data()), data.size());
+#elif TORRENT_USE_CRYPTOAPI
+		if (CryptHashData(m_context, reinterpret_cast<BYTE const*>(data.data()), int(data.size()), 0) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+#elif defined TORRENT_USE_LIBCRYPTO
+		SHA512_Update(&m_context, reinterpret_cast<unsigned char const*>(data.data()), data.size());
+#else
+		SHA512_update(&m_context, reinterpret_cast<unsigned char const*>(data.data()), data.size());
+#endif
+		return *this;
+	}
+
+	void hasher512::final(span<char> digest)
```

Comment:
well, I don't like it either, I just did't figure out the best alternative. Let me see what can I do with `sha1_hash`